tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java

```diff
@@ -0,0 +1,376 @@
+        private int partition;
+        private Long timestamp;
+
+        public Row(String name, int partition, Long timestamp) {
+            this.name = name;
+            this.partition = partition;
+            this.timestamp = timestamp;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (o == this) return true;
+
+            if (!(o instanceof Row)) return false;
+
+            Row r = (Row) o;
+
+            return name.equals(r.name) && partition == r.partition && Objects.equals(timestamp, r.timestamp);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(name, partition, timestamp);
+        }
+    }
+
+    @ClusterTest
+    public void testNoFilterOptions() {
+        setUp();
+
+        List<Row> output = executeAndParse();
+
+        assertEquals(expectedOffsetsWithInternal(), output);
+    }
+
+    @ClusterTest
+    public void testInternalExcluded() {
+        setUp();
+
+        List<Row> output = executeAndParse("--exclude-internal-topics");
+
+        assertEquals(expectedTestTopicOffsets(), output);
+    }
+
+    @ClusterTest
+    public void testTopicNameArg() {
+        setUp();
+
+        IntStream.range(1, topicCount + 1).forEach(i -> {
+            List<Row> offsets = executeAndParse("--topic", getTopicName(i));
+
+            assertEquals(expectedOffsetsForTopic(i), offsets, () -> "Offset output did not match for " + getTopicName(i));
+        });
+    }
+
+    @ClusterTest
+    public void testTopicPatternArg() {
+        setUp();
+
+        List<Row> offsets = executeAndParse("--topic", "topic.*");
+
+        assertEquals(expectedTestTopicOffsets(), offsets);
+    }
+
+    @ClusterTest
+    public void testPartitionsArg() {
+        setUp();
+
+        List<Row> offsets = executeAndParse("--partitions", "0,1");
+
+        assertEquals(expectedOffsetsWithInternal().stream().filter(r -> r.partition <= 1).collect(Collectors.toList()), offsets);
+    }
+
+    @ClusterTest
+    public void testTopicPatternArgWithPartitionsArg() {
+        setUp();
+
+        List<Row> offsets = executeAndParse("--topic", "topic.*", "--partitions", "0,1");
+
+        assertEquals(expectedTestTopicOffsets().stream().filter(r -> r.partition <= 1).collect(Collectors.toList()), offsets);
+    }
+
+    @ClusterTest
+    public void testTopicPartitionsArg() {
+        setUp();
+
+        List<Row> offsets = executeAndParse("--topic-partitions", "topic1:0,topic2:1,topic(3|4):2,__.*:3");
+        List<Row> expected = Arrays.asList(
+                new Row("__consumer_offsets", 3, 0L),
+                new Row("topic1", 0, 1L),
+                new Row("topic2", 1, 2L),
+                new Row("topic3", 2, 3L),
+                new Row("topic4", 2, 4L)
+        );
+
+        assertEquals(expected, offsets);
+    }
+
+    @ClusterTest
+    public void testGetLatestOffsets() {
+        setUp();
+
+        for (String time : new String[] {"-1", "latest"}) {
```

Comment:
I also didn't find a way how to use `ParametrizedTest` and `ValueSource` in the same place with `ClusterTest`. It throws error about `ParameterResolver`, so I did this approach.